### ISSUE TITLE
cmd/go: build constraints sometimes ineffective for net/cgo_resnew.go on FreeBSD

### ISSUE THREAD

**Issue by ohmantics** [golang/go#33101]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.12.6 freebsd/amd64
</pre>

### Does this issue reproduce with the latest release?
Yes. (Bumped to 1.12.7 by changing pkg repos over to latest and tried again.)

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/xxx/.cache/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="freebsd"
GOOS="freebsd"
GOPATH="/home/xxx/go"
GOPROXY=""
GORACE=""
GOROOT="/usr/local/go"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/freebsd_amd64"
GCCGO="gccgo"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/home/xxx/Projects/sshcode/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build447432705=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

### What did you do?
Downloaded [sshcode](https://github.com/cdr/sshcode) and was in progress patching it for FreeBSD. Ran `ci/build.sh`.

### What did you expect to see?
Build success.

### What did you see instead?
A few repeats of the below:

```
# net
/usr/local/go/src/net/cgo_resnew.go:21:87: cannot use _Ctype_uint(len(b)) (type _Ctype_uint) as type _Ctype_ulong in argument to _C2func_getnameinfo
```

**Comment by bcmills** [comment]
The [`+build` constraints](https://github.com/golang/go/blob/89d300b821551c9955414caf3066ae602813f6ea/src/net/cgo_resnew.go#L5-L6) on `cgo_resnew.go` do not appear to allow `freebsd`: it should be using `cgo_resold.go` instead.

Can you reproduce this failure using a sequence of `go` commands rather than the `ci/build.sh` wrapper?

**Comment by bcmills** [comment]
Ah, yeah: `ci/build.sh` is explicitly [cross-compiling](https://github.com/cdr/sshcode/blob/1b15b2be9b051665ac26c6e14e67f453e6e700f3/ci/build.sh#L20-L21), so presumably you don't have the correct C headers for the target platforms.

Cross-compiles don't enable cgo by default, for that exact reason. Do you have `CGO_ENABLED` explicitly set in your environment? If so, what happens if you unset it?

**Comment by cuonglm** [comment]
Same errors in build log https://storage.googleapis.com/go-build-log/9e20f242/freebsd-amd64-12_0_3ca8deda.log

**Comment by bcmills** [comment]
@cuonglm, was that a flake, or did you find a workaround?

**Comment by cuonglm** [comment]
@bcmills it does not seem to be a flake, it occrurred when I cross-compile with cgo enable on freebsd.

In my CL, I change the test to run on Linux only.

**Comment by stevenh** [comment]
Have the same problem here building one of our packages, confirmed on:
`go version go1.12.9 freebsd/amd64`

Setting CGO_ENABLED=0 fixed it but I thought cross-compile did this by default?

**Comment by bcmills** [comment]
@stevenh, @cuonglm, if you can reproduce this reliably please post exact instructions, including the output of `go env` and `uname -sv`.

**Comment by bcmills** [comment]
I can reproduce the observed symptom on FreeBSD 12.1 by running:
```
CGO_ENABLED=1 GOOS=linux GOARCH=amd64 go build net
```

Without the explicit `CGO_ENABLED=1`, that command succeeds.

I note that `sshcode/build.sh` explicitly sets `GOOS=linux`, so all that would be required to reproduce it there is `CGO_ENABLED=1`.

So my question to all of you is: have you seen this issue with `CGO_ENABLED` unset in the environment of the `go` command in which you observed the failure?

**Comment by stevenh** [comment]
With `1.12.9` I get
```
go build
# net
/usr/local/go/src/net/cgo_resnew.go:21:87: cannot use _Ctype_uint(len(b)) (type _Ctype_uint) as type _Ctype_ulong in argument to _C2func_getnameinfo
```

```
go env
GOARCH="amd64"
GOBIN="/data/go/bin"
GOCACHE="/root/.cache/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="freebsd"
GOOS="linux"
GOPATH="/data/go"
GOPROXY=""
GORACE=""
GOROOT="/usr/local/go"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/freebsd_amd64"
GCCGO="gccgo"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build371148949=/tmp/go-build -gno-record-gcc-switches"
```
```
uname -sv
FreeBSD FreeBSD 12.0-RELEASE-p10 MULTIPLAY-r5
```
```
env |grep GO
GOPATH=/data/go
GOHOME=/data/go
GOBIN=/data/go/bin
GOOS=linux
```

With  `1.13` it doesn't exhibit the same behaviour, and notably `go env` has changed to `CGO_ENABLED="0"` which under `1.12.9` was `CGO_ENABLED="1"` even though its *not* set in my environment.

```
go env
GO111MODULE=""
GOARCH="amd64"
GOBIN="/data/go/bin"
GOCACHE="/root/.cache/go-build"
GOENV="/root/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="freebsd"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/data/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/freebsd_amd64"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="0"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build722621377=/tmp/go-build -gno-record-gcc-switches"
```

In addition removing `GOOS=linux` from my env makes `1.13` automatically set `CGO_ENABLED="1"` so it looks like `1.12` had an issue correctly configuring `CGO_ENABLED` when cross compiling.

So to confirm looks like this is fixed in `1.13`.

**Comment by bcmills** [comment]
Thanks for the followup.

#33673 seems closely related. The report there suggests that the default value of `CGO_ENABLED` for cross-compiled builds may have changed back to `0` (as intended) in [CL 171137](https://golang.org/cl/171137).

So I'm going to call this “fixed in 1.13”, but please open a new issue if it crops up again!
